File: src/main/java/ltdjms/discord/gametoken/services/DiceGame1Service.java
package ltdjms.discord.gametoken.services;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.currency.domain.CurrencyTransaction;
import ltdjms.discord.currency.domain.MemberCurrencyAccount;
import ltdjms.discord.currency.persistence.MemberCurrencyAccountRepository;
import ltdjms.discord.currency.services.CurrencyTransactionService;
import ltdjms.discord.gametoken.domain.DiceGame1Config;
import ltdjms.discord.shared.events.BalanceChangedEvent;
import ltdjms.discord.shared.events.DomainEventPublisher;

/**
 * Service for the dice-game-1 mini-game. Handles dice rolling, reward calculation, and currency
 * distribution.
 *
 * <p>The number of dice rolled is determined by the number of tokens spent: 1 token = 1 dice.
 */
public class DiceGame1Service {

  private static final Logger LOG = LoggerFactory.getLogger(DiceGame1Service.class);

  private final MemberCurrencyAccountRepository currencyRepository;
  private final CurrencyTransactionService transactionService;
  private final DomainEventPublisher eventPublisher;
  private final Random random;

  public DiceGame1Service(
      MemberCurrencyAccountRepository currencyRepository,
      CurrencyTransactionService transactionService,
      DomainEventPublisher eventPublisher) {
    this(currencyRepository, transactionService, eventPublisher, new Random());
  }

  /** Constructor with injectable Random for testing. */
  public DiceGame1Service(
      MemberCurrencyAccountRepository currencyRepository,
      CurrencyTransactionService transactionService,
      DomainEventPublisher eventPublisher,
      Random random) {
    this.currencyRepository = currencyRepository;
    this.transactionService = transactionService;
    this.eventPublisher = eventPublisher;
    this.random = random;
  }

  /**
   * Plays the dice game for a member using the provided configuration and token amount. Rolls one
   * dice per token spent and calculates the total reward based on dice values. The reward is added
   * to the member's currency account.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param config the game configuration containing reward multiplier
   * @param diceCount the number of dice to roll (equals tokens spent)
   * @return the game result
   */
  public DiceGameResult play(long guildId, long userId, DiceGame1Config config, int diceCount) {
    LOG.debug(
        "Playing dice-game-1 for guildId={}, userId={}, diceCount={}, rewardPerDice={}",
        guildId,
        userId,
        diceCount,
        config.rewardPerDiceValue());

    // Roll dice
    List<Integer> diceRolls = rollDice(diceCount);

    // Calculate total reward using configured multiplier
    long totalReward = calculateTotalReward(diceRolls, config.rewardPerDiceValue());

    // Apply reward to currency account (may need multiple adjustments due to MAX_ADJUSTMENT_AMOUNT)
    long previousBalance = currencyRepository.findOrCreate(guildId, userId).balance();
    applyRewardToCurrency(guildId, userId, totalReward);
    long newBalance =
        currencyRepository
            .findByGuildIdAndUserId(guildId, userId)
            .map(MemberCurrencyAccount::balance)
            .orElse(previousBalance + totalReward);

    // Record transaction
    if (totalReward > 0) {
      transactionService.recordTransaction(
          guildId,
          userId,
          totalReward,
          newBalance,
          CurrencyTransaction.Source.DICE_GAME_1_WIN,
          null);
    }

    // Publish event
    eventPublisher.publish(new BalanceChangedEvent(guildId, userId, newBalance));

    DiceGameResult result =
        new DiceGameResult(guildId, userId, diceRolls, totalReward, previousBalance, newBalance);

    LOG.info(
        "Dice game completed: guildId={}, userId={}, rolls={}, reward={}, newBalance={}",
        guildId,
        userId,
        diceRolls,
        totalReward,
        newBalance);

    return result;
  }

  /**
   * Rolls the specified number of dice.
   *
   * @param count the number of dice to roll
   * @return list of dice values (1-6)
   */
  List<Integer> rollDice(int count) {
    List<Integer> rolls = new ArrayList<>(count);
    for (int i = 0; i < count; i++) {
      rolls.add(random.nextInt(6) + 1); // 1-6
    }
    return rolls;
  }

  /**
   * Calculates the total reward based on dice rolls and configured reward multiplier.
   *
   * @param diceRolls the list of dice values
   * @param rewardPerDiceValue the reward multiplier per dice value
   * @return the total reward
   */
  long calculateTotalReward(List<Integer> diceRolls, long rewardPerDiceValue) {
    long sum = 0;
    for (int i = 0; i < diceRolls.size(); i++) {
      sum += diceRolls.get(i);
    }
    return sum * rewardPerDiceValue;
  }

  /**
   * Applies the reward to the member's currency account. If the reward exceeds the max adjustment
   * amount, splits into multiple adjustments.
   *
   * @param guildId the Discord guild ID
   * @param userId the Discord user ID
   * @param totalReward the total reward to apply
   */
  void applyRewardToCurrency(long guildId, long userId, long totalReward) {
    long remaining = totalReward;
    long maxAdjustment = MemberCurrencyAccount.MAX_ADJUSTMENT_AMOUNT;

    while (remaining > 0) {
      long adjustment = Math.min(remaining, maxAdjustment);
      currencyRepository.adjustBalance(guildId, userId, adjustment);
      remaining -= adjustment;
    }
  }

  /** Result of a dice game. */
  public record DiceGameResult(
      long guildId,
      long userId,
      List<Integer> diceRolls,
      long totalReward,
      long previousBalance,
      long newBalance) {
    /** Formats the result as a Discord message. */
    public String formatMessage(String currencyIcon, String currencyName) {
      StringBuilder sb = new StringBuilder();
      sb.append("**Dice Game Results**\n");
      sb.append("Rolls: ");

      for (int i = 0; i < diceRolls.size(); i++) {
        int roll = diceRolls.get(i);
        sb.append(diceEmoji(roll));
        if (i < diceRolls.size() - 1) {
          sb.append(" ");
        }
      }

      sb.append("\n\n");
      sb.append(
          String.format("Total Reward: %s %,d %s\n", currencyIcon, totalReward, currencyName));
      sb.append(String.format("New Balance: %s %,d %s", currencyIcon, newBalance, currencyName));

      return sb.toString();
    }

    private String diceEmoji(int value) {
      return switch (value) {
        case 1 -> ":one:";
        case 2 -> ":two:";
        case 3 -> ":three:";
        case 4 -> ":four:";
        case 5 -> ":five:";
        case 6 -> ":six:";
        default -> String.valueOf(value);
      };
    }
  }
}

File: packages/games/src/dice/services/dice-game-1-service.ts
import { type Result, Ok, Err, DomainError } from '@ltdjms/shared';
import { GameRewardService } from './game-reward-service.js';
import type { DiceGame1Config, DiceGame1Result } from '../../domain/types.js';
import { CurrencyTransactionSource } from '@ltdjms/economy';
import { type Random, DefaultRandom, rollDice as randomRollDice } from './random.js';

/**
 * Rolls dice deterministically (for testing with predetermined values).
 * @internal Exposed for test use only; not part of the public API.
 */
export function rollDice(count: number, random: Random): number[] {
  return randomRollDice(count, random);
}

/**
 * Calculates the total reward from dice rolls.
 * @internal Exposed for test use only; not part of the public API.
 */
export function calculateTotalReward(
  diceRolls: readonly number[],
  rewardPerDiceValue: number,
): number {
  const sum = diceRolls.reduce((acc, val) => acc + val, 0);
  return sum * rewardPerDiceValue;
}

/**
 * Dice Game 1 service implementation.
 * One token = one die rolled. Reward = sum(dice) * rewardPerDiceValue.
 * Matches Java DefaultDiceGame1Service behavior exactly.
 *
 * This service focuses on game logic only: rolling dice and calculating rewards.
 * Config lookup and token deduction are handled by the command handler.
 */
export class DiceGame1Service {
  constructor(
    private readonly gameRewardService: GameRewardService,
    private readonly random: Random = DefaultRandom,
  ) {}

  /**
   * Plays the dice game. Rolls dice and calculates reward.
   * The number of dice equals the number of tokens spent.
   * Config validation and token deduction must be done by the caller (handler).
   */
  async play(
    guildId: number,
    userId: string,
    tokenCount: number,
    config: DiceGame1Config,
  ): Promise<Result<DiceGame1Result, DomainError>> {
    // Validate token count against config
    if (tokenCount < config.minTokensPerPlay) {
      return new Err(
        DomainError.invalidInput(
          `Token count ${tokenCount} is less than minimum ${config.minTokensPerPlay}`,
        ),
      );
    }
    if (tokenCount > config.maxTokensPerPlay) {
      return new Err(
        DomainError.invalidInput(
          `Token count ${tokenCount} exceeds maximum ${config.maxTokensPerPlay}`,
        ),
      );
    }

    // Roll dice - one die per token
    const diceRolls: number[] = [];
    for (let i = 0; i < tokenCount; i++) {
      // random.nextInt(6) + 1 gives values 1-6 (matching Java''s Random.nextInt(6) + 1)
      diceRolls.push(this.random.nextInt(6) + 1);
    }

    // Calculate total reward: sum(dice) * rewardPerDiceValue
    const diceSum = diceRolls.reduce((acc, val) => acc + val, 0);
    const totalReward = diceSum * config.rewardPerDiceValue;

    // Apply reward via GameRewardService — returns both previous and new balance (P1-25)
    const rewardResult = await this.gameRewardService.creditReward(
      guildId,
      userId,
      totalReward,
      CurrencyTransactionSource.DICE_GAME_1_WIN,
    );

    if (rewardResult.isErr()) {
      return new Err(rewardResult.getError());
    }
    const { previousBalance, newBalance } = rewardResult.getValue();

    return new Ok({
      guildId,
      userId,
      diceRolls,
      diceSum,
      totalReward,
      previousBalance,
      newBalance,
    });
  }
}
